analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CloudNative/AzureFunctionsLogFailures.cs

```diff
@@ -0,0 +1,325 @@
+//          ^^^^^^^^^^^^^^^^^^^^^^^^^^
+            log.LogDebug(string.Empty);        // Secondary
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionInCatchClause(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // Compliant
+        {
+            log.LogError(ex, "");
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionInWrappedLogger(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // Compliant
+        {
+            var nullLogger = NullLogger.Instance;
+            nullLogger.Log(LogLevel.Error, new EventId(), (object)null, ex, (s, e) => string.Empty);
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionLoggedInLocalFunction(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // FN. Any call to "log" is considered valid. Reachability is not considered.
+        {
+            void Log() => log.LogError("");
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionLoggedInLambda(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // FN. Any call to "log" is considered valid. Reachability is not considered.
+        {
+            Action x = () => log.LogError("");
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionLoggedInAnonymousFunction(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // FN. Any call to "log" is considered valid. Reachability is not considered.
+        {
+            Action x = delegate () { log.LogError(""); };
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionIsPassedInLambda(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // FN. Any call to "log" is considered valid. Reachability is not considered.
+        {
+            DoSomething(() => log.LogError(""));
+        }
+
+        void DoSomething(Action a) => a();
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionLoggedInUnreachableCode(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // FN. log in unreachable code.
+        {
+            return;
+            log.LogError("");
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionNotEveryCodePathContainsLog(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // FN. Not every code path contains a log
+        {
+            var i = 100;
+            if (i == 42)
+            {
+                log.LogError("");
+            }
+        }
+    }
+
+    [FunctionName("Sample")]
+    public static void LogExceptionFromWebApi(ILogger log)
+    {
+        try { }
+        catch (Exception ex) // Compliant. Call to unregcognized extension method from WebApi
+        {
+            var i = 100;
+            if (i == 42)
+            {
+                log.LogMetric("Metric", 5);
```

Comment:
`LogMetric` is something that doesn't exist and returns a `null` symbol? I'd be explicit about that, because I don't know the interface:
```suggestion
                log.UndefinedMethod("Metric", 5);
```

Reply:
If `LogMetric` is defined, I'd ask why is it compliant then.

Reply:
The `WebApi` comment will need to be updated in case it should be undefined. Because then it will not matter where it comes from.

Reply:
It comes from microsoft.azure.webjobs and is part of the scaffolded AzureFunctions solution. We don't explicit recognize it, so it is treated as any other unknown call to ILogger (and therefore it is compliant).

Reply:
I added one more test with an undefined method, though.